python/ctsm/utils.py

```diff
@@ -33,3 +46,149 @@ def fill_template_file(path_to_template, path_to_final, substitutions):
+
+    for attr in del_attrs:
+        if attr in attr_list:
+            del file.attrs[attr]
+
+def lon_range_0_to_360(lon_in):
+    """
+    Description
+    -----------
+    Restrict longitude to 0 to 360 when given as -180 to 180.
+    """
+    if -180 <= lon_in < 0:
+        lon_out = lon_in + 360
+        message = 'INFO: Resetting longitude from ' + str(lon_in) + \
+                  ' to ' + str(lon_out) + ' to keep in the range 0 to 360'
+        print(message)  # TODO Use logging to print this
+    elif 0 <= lon_in <= 360 or lon_in is None:
+        lon_out = lon_in
+    else:
+        errmsg = 'lon_in needs to be in the range 0 to 360'
+        abort(errmsg)
+
+    return lon_out
+
+def get_config_value(config, section, item, file_path, allowed_values=None, default=None, is_list=False, convert_to_type=None):
+    """Get a given item from a given section of the config object
+    Give a helpful error message if we can't find the given section or item
+    Note that the file_path argument is only used for the sake of the error message
+    If allowed_values is present, it should be a list of strings giving allowed values
+    """
+    try:
+        val = config.get(section, item)
+    except NoSectionError:
+        abort("ERROR: Config file {} must contain section '{}'".format(file_path, section))
+    except NoOptionError:
+        abort("ERROR: Config file {} must contain item '{}' in section '{}'".format(
+            file_path, item, section))
+
+    if val == _CONFIG_PLACEHOLDER:
+        abort("Error: {} needs to be specified in config file {}".format(item, file_path))
+
+    if allowed_values is not None:
+        if val not in allowed_values:
+            abort("Error: {} is not an allowed value for {} in config file {}\n"
+                  "Allowed values: {}".format(val, item, file_path, allowed_values))
+
+    val = _handle_config_value(var=val, default=default, item=item,
+        is_list=is_list, convert_to_type=convert_to_type)
+
+    return val
+
+def _handle_config_value(var, default, item, is_list, convert_to_type):
+    """
+    Description
+    -----------
+    Assign the default value or the user-specified one to var.
+    Convert from default type (str) to reqested type (int or float).
+    """
+    if var == CONFIG_UNSET:
+        var = default  # default may be None
+
+    if var is not None and is_list:
+        var = list(var.split())  # convert string to list of strings
+        try:
+            var = list(map(convert_to_type, var))  # convert
+        except Exception as err:
+            errmsg = '--> Check .cfg file variable: ' + item
+            err.args = (str(err) + errmsg,)
+            raise
+
+    if var is not None and not is_list and convert_to_type is not None:
+        try:
+            var = convert_to_type(var)
+        except Exception as err:
+            errmsg = '--> Check .cfg file variable: ' + item
+            err.args = (str(err) + errmsg,)
+            raise
```

Comment:
- [ ] The `is_list` block assumes that `convert_to_type` is not None. I think you could consolidate this code as follows (untested); this also implements a number of my other suggestions:

``` python
def _handle_config_value(var, default, item, is_list, convert_to_type, can_be_unset, allowed_values):
    """
    Description
    -----------
    Assign the default value or the user-specified one to var.
    Convert from default type (str) to reqested type (int or float).

    If is_list is True, then default should be a list
    """
    if var == CONFIG_UNSET:
        if can_be_unset:
            return default  # default may be None
        abort('Must set a value for .cfg file variable: {}'.format(item))

    # convert string to list of strings; if there is just one element, we will get a list
    # of size one, which we will convert back to a scalar later if needed
    var = var.split()

    if convert_to_type is bool:
        try:
            var = [_convert_to_bool(v) for v in var]
        except ValueError:
            abort('Non-boolean value found for .cfg file variable: {}'.format(item))
    elif convert_to_type is not None:
        try:
            var = [convert_to_type(v) for v in var]
        except ValueError:
            abort('Wrong type for .cfg file variable: {}'.format(item))

    if allowed_values is not None:
        for v in var:
            if v not in allowed_values:
                abort("{} is not an allowed value for {} in .cfg file\n"
                      "Allowed values: {}".format(v, item, allowed_values))

    if not is_list:
        if len(var) > 1:
            abort('More than 1 element found for .cfg file variable: {}'.format(item))
        var = var[0]

    return var

def _convert_to_bool(val):
    """Convert the given value to boolean

    Conversion is as in config files 'getboolean'
    """
    if val.lower() in ['1', 'yes', 'true', 'on']:
        return True
    if val.lower() in ['0', 'no', 'false', 'off']:
        return False
    raise ValueError("{} cannot be converted to boolean".format(val))
```

- [ ] Note that I have added the `can_be_unset` handling that I suggested in an earlier comment. You should therefore remove 'CONFIG_UNSET' as an allowed value for some variables, setting `can_be_unset` to `True` for those.

- [ ] Note that I have moved the allowed value checking into here; it can then be removed from `get_config_value`

A few other notes about this:
- In the initial check for `CONFIG_UNSET`, I return immediately. I think that should work: I don't think we should need any of the later processing in this case (and some of the later processing may not work right if we try to run it after setting a default value). That also  removes the need for checking for `None` values later in the code (I assume those checks were just to handle the possibility that we had set the value to the default, which was `None`).
- I have changed `list(var.split())` to simply `var.split()`: I think this already returns a list so the conversion to list is unnecessary (unless you think otherwise)
- I have changed `list(map(convert_to_type, var))` to the equivalent list comprehension, `[convert_to_type(v) for v in var]`. They are equivalent, but I personally find the list comprehension easier to read. I'm really fine with either, though: my main point was to illustrate how you could do this with a list comprehension in case you aren't as familiar with that form; if you prefer the `map`, then by all means stick with it.
- I have changed the exception handling to call abort rather than raising an exception, because the latter can be ugly for users to have to read through; you should try this and make sure the error message looks good
- Note that I have implemented my suggested conversion-to-boolean function here (see an earlier comment in this review related to this)

- [ ] Finally, there should be some unit tests of both of these functions (`_convert_to_bool` and `_handle_config_value`)

Reply:
As of now I have addressed this **except for the unit tests.**
(I hope I didn't miss any details, but that's what the unit tests will help with.)

Reply:
@billsacks as always thank you for the well-written code and clear instructions by the way!